# Homicide Dataset

## Loading and clean up

In [70]:
import numpy as np
import pandas as pd

In [71]:
homicide = pd.read_csv('homicide-reports.csv')
homicide.head()

/Users/ctoon/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (16) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Record ID Agency Code Agency Name       Agency Type       City   State  \
0          1     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
1          2     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
2          3     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
3          4     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
4          5     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   

   Year    Month  Incident              Crime Type      ...       \
0  1980  January         1  Murder or Manslaughter      ...        
1  1980    March         1  Murder or Manslaughter      ...        
2  1980    March         2  Murder or Manslaughter      ...        
3  1980    April         1  Murder or Manslaughter      ...        
4  1980    April         2  Murder or Manslaughter      ...        

  Victim Ethnicity Perpetrator Sex  Perpetrator Age  \
0          Unknown            Male               15   
1          Unknown            Male               42   
2          Unknown         Unknown                0   
3          Unknown            Male               42   
4          Unknown         Unknown                0   

                Perpetrator Race Perpetrator Ethnicity  Relationship  \
0  Native American/Alaska Native               Unknown  Acquaintance   
1                          White               Unknown  Acquaintance   
2                        Unknown               Unknown       Unknown   
3                          White               Unknown  Acquaintance   
4                        Unknown               Unknown       Unknown   

          Weapon Victim Count Perpetrator Count Record Source  
0   Blunt Object            0                 0           FBI  
1  Strangulation            0                 0           FBI  
2        Unknown            0                 0           FBI  
3  Strangulation            0                 0           FBI  
4        Unknown            0                 1           FBI  

[5 rows x 24 columns]

An error was thrown indicating column 16 (Perpetrator Age) contains mixed type. To resolve this, I'll convert the entire column to numeric and coerce non numeric values to NaN.

In [72]:
homicide['Perpetrator Age'] = pd.to_numeric(homicide['Perpetrator Age'], errors='coerce')
print(homicide['Perpetrator Age'].dtype)

float64


I actually want ints, not floats. I want to convert floats to int, but cannot because there is a null value somewhere.

In [73]:
print('Perpetrator Age:', homicide['Perpetrator Age'].isnull().any())
print('Homicide DF:', homicide.isnull().any())

Perpetrator Age: True
Homicide DF: Record ID                False
Agency Code              False
Agency Name              False
Agency Type              False
City                     False
State                    False
Year                     False
Month                    False
Incident                 False
Crime Type               False
Crime Solved             False
Victim Sex               False
Victim Age               False
Victim Race              False
Victim Ethnicity         False
Perpetrator Sex          False
Perpetrator Age           True
Perpetrator Race         False
Perpetrator Ethnicity    False
Relationship             False
Weapon                   False
Victim Count             False
Perpetrator Count        False
Record Source            False
dtype: bool


Only Perpetrator Age has a null value

In [74]:
homicide.dropna(axis=0, how='any', inplace=True)
homicide.isnull().any()

Record ID                False
Agency Code              False
Agency Name              False
Agency Type              False
City                     False
State                    False
Year                     False
Month                    False
Incident                 False
Crime Type               False
Crime Solved             False
Victim Sex               False
Victim Age               False
Victim Race              False
Victim Ethnicity         False
Perpetrator Sex          False
Perpetrator Age          False
Perpetrator Race         False
Perpetrator Ethnicity    False
Relationship             False
Weapon                   False
Victim Count             False
Perpetrator Count        False
Record Source            False
dtype: bool

In [75]:
homicide['Perpetrator Age'] = homicide['Perpetrator Age'].astype(int)
homicide['Perpetrator Age'].dtype

dtype('int64')

In [76]:
homicide.head()

Record ID Agency Code Agency Name       Agency Type       City   State  \
0          1     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
1          2     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
2          3     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
3          4     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   
4          5     AK00101   Anchorage  Municipal Police  Anchorage  Alaska   

   Year    Month  Incident              Crime Type      ...       \
0  1980  January         1  Murder or Manslaughter      ...        
1  1980    March         1  Murder or Manslaughter      ...        
2  1980    March         2  Murder or Manslaughter      ...        
3  1980    April         1  Murder or Manslaughter      ...        
4  1980    April         2  Murder or Manslaughter      ...        

  Victim Ethnicity Perpetrator Sex  Perpetrator Age  \
0          Unknown            Male               15   
1          Unknown            Male               42   
2          Unknown         Unknown                0   
3          Unknown            Male               42   
4          Unknown         Unknown                0   

                Perpetrator Race Perpetrator Ethnicity  Relationship  \
0  Native American/Alaska Native               Unknown  Acquaintance   
1                          White               Unknown  Acquaintance   
2                        Unknown               Unknown       Unknown   
3                          White               Unknown  Acquaintance   
4                        Unknown               Unknown       Unknown   

          Weapon Victim Count Perpetrator Count Record Source  
0   Blunt Object            0                 0           FBI  
1  Strangulation            0                 0           FBI  
2        Unknown            0                 0           FBI  
3  Strangulation            0                 0           FBI  
4        Unknown            0                 1           FBI  

[5 rows x 24 columns]

In [77]:
homicide.shape

(638453, 24)

## Exploration

In [78]:
ma_homicides = homicide[homicide['State'] == 'Massachusetts']
print(ma_homicides.shape)
ma_homicides.head()

(6036, 24)


Record ID Agency Code Agency Name       Agency Type        City  \
10230      10231     MA00102      Bourne  Municipal Police  Barnstable   
10231      10232     MA00102      Bourne  Municipal Police  Barnstable   
10232      10233     MA00109     Mashpee  Municipal Police  Barnstable   
10233      10234     MA00201       Adams  Municipal Police   Berkshire   
10234      10235     MA00212   Lanesboro  Municipal Police   Berkshire   

               State  Year      Month  Incident              Crime Type  \
10230  Massachusetts  1980    January         1  Murder or Manslaughter   
10231  Massachusetts  1980     August         1  Murder or Manslaughter   
10232  Massachusetts  1980  September         1  Murder or Manslaughter   
10233  Massachusetts  1980     August         1  Murder or Manslaughter   
10234  Massachusetts  1980       June         1  Murder or Manslaughter   

           ...      Victim Ethnicity Perpetrator Sex  Perpetrator Age  \
10230      ...               Unknown         Unknown                0   
10231      ...          Not Hispanic            Male               15   
10232      ...          Not Hispanic            Male               21   
10233      ...          Not Hispanic            Male               22   
10234      ...          Not Hispanic         Unknown                0   

      Perpetrator Race Perpetrator Ethnicity  Relationship         Weapon  \
10230          Unknown               Unknown       Unknown          Knife   
10231            White          Not Hispanic        Family   Blunt Object   
10232            White          Not Hispanic  Acquaintance   Blunt Object   
10233            White          Not Hispanic      Stranger   Blunt Object   
10234          Unknown               Unknown       Unknown  Strangulation   

      Victim Count Perpetrator Count Record Source  
10230            0                 1           FBI  
10231            0                 0           FBI  
10232            0                 1           FBI  
10233            0                 1           FBI  
10234            0                 0           FBI  

[5 rows x 24 columns]

In [79]:
print(homicide.columns)
len(homicide.columns)

Index(['Record ID', 'Agency Code', 'Agency Name', 'Agency Type', 'City',
       'State', 'Year', 'Month', 'Incident', 'Crime Type', 'Crime Solved',
       'Victim Sex', 'Victim Age', 'Victim Race', 'Victim Ethnicity',
       'Perpetrator Sex', 'Perpetrator Age', 'Perpetrator Race',
       'Perpetrator Ethnicity', 'Relationship', 'Weapon', 'Victim Count',
       'Perpetrator Count', 'Record Source'],
      dtype='object')


24

In [80]:
homicide = homicide.iloc[:,4:24]
homicide.head()

City   State  Year    Month  Incident              Crime Type  \
0  Anchorage  Alaska  1980  January         1  Murder or Manslaughter   
1  Anchorage  Alaska  1980    March         1  Murder or Manslaughter   
2  Anchorage  Alaska  1980    March         2  Murder or Manslaughter   
3  Anchorage  Alaska  1980    April         1  Murder or Manslaughter   
4  Anchorage  Alaska  1980    April         2  Murder or Manslaughter   

  Crime Solved Victim Sex  Victim Age                    Victim Race  \
0          Yes       Male          14  Native American/Alaska Native   
1          Yes       Male          43                          White   
2           No     Female          30  Native American/Alaska Native   
3          Yes       Male          43                          White   
4           No     Female          30  Native American/Alaska Native   

  Victim Ethnicity Perpetrator Sex  Perpetrator Age  \
0          Unknown            Male               15   
1          Unknown            Male               42   
2          Unknown         Unknown                0   
3          Unknown            Male               42   
4          Unknown         Unknown                0   

                Perpetrator Race Perpetrator Ethnicity  Relationship  \
0  Native American/Alaska Native               Unknown  Acquaintance   
1                          White               Unknown  Acquaintance   
2                        Unknown               Unknown       Unknown   
3                          White               Unknown  Acquaintance   
4                        Unknown               Unknown       Unknown   

          Weapon  Victim Count  Perpetrator Count Record Source  
0   Blunt Object             0                  0           FBI  
1  Strangulation             0                  0           FBI  
2        Unknown             0                  0           FBI  
3  Strangulation             0                  0           FBI  
4        Unknown             0                  1           FBI

In [81]:
columns = homicide.columns
columns

Index(['City', 'State', 'Year', 'Month', 'Incident', 'Crime Type',
       'Crime Solved', 'Victim Sex', 'Victim Age', 'Victim Race',
       'Victim Ethnicity', 'Perpetrator Sex', 'Perpetrator Age',
       'Perpetrator Race', 'Perpetrator Ethnicity', 'Relationship', 'Weapon',
       'Victim Count', 'Perpetrator Count', 'Record Source'],
      dtype='object')

In [82]:
homicide['Crime Type'].value_counts()

Murder or Manslaughter        629337
Manslaughter by Negligence      9116
Name: Crime Type, dtype: int64

In [83]:
homicide['Crime Solved'].value_counts()

Yes    448172
No     190281
Name: Crime Solved, dtype: int64

In [84]:
# Percent of unsolved crimes
print(190281 / (190281 + 448172))

0.29803446768986913


## Exploring unsolved crimes

In [85]:
unsolved = homicide[homicide['Crime Solved'] == 'No']
print(unsolved.shape)
unsolved.head()

(190281, 20)


City   State  Year   Month  Incident              Crime Type  \
2   Anchorage  Alaska  1980   March         2  Murder or Manslaughter   
4   Anchorage  Alaska  1980   April         2  Murder or Manslaughter   
8   Anchorage  Alaska  1980    June         2  Murder or Manslaughter   
10  Anchorage  Alaska  1980    July         1  Murder or Manslaughter   
14  Anchorage  Alaska  1980  August         2  Murder or Manslaughter   

   Crime Solved Victim Sex  Victim Age                    Victim Race  \
2            No     Female          30  Native American/Alaska Native   
4            No     Female          30  Native American/Alaska Native   
8            No       Male          32                          White   
10           No       Male          36  Native American/Alaska Native   
14           No       Male          48                          White   

   Victim Ethnicity Perpetrator Sex  Perpetrator Age Perpetrator Race  \
2           Unknown         Unknown                0          Unknown   
4           Unknown         Unknown                0          Unknown   
8           Unknown         Unknown                0          Unknown   
10          Unknown         Unknown                0          Unknown   
14          Unknown         Unknown                0          Unknown   

   Perpetrator Ethnicity Relationship   Weapon  Victim Count  \
2                Unknown      Unknown  Unknown             0   
4                Unknown      Unknown  Unknown             0   
8                Unknown      Unknown  Firearm             0   
10               Unknown      Unknown  Unknown             0   
14               Unknown      Unknown  Handgun             0   

    Perpetrator Count Record Source  
2                   0           FBI  
4                   1           FBI  
8                   0           FBI  
10                  1           FBI  
14                  0           FBI

In [86]:
unsolved['Victim Ethnicity'].value_counts()

Unknown         113271
Not Hispanic     51918
Hispanic         25092
Name: Victim Ethnicity, dtype: int64

In [87]:
unsolved['Weapon'].value_counts()

Handgun          100545
Firearm           22578
Knife             20204
Blunt Object      15833
Unknown           14783
Shotgun            4682
Strangulation      3701
Rifle              3401
Fire               2034
Suffocation         990
Gun                 882
Drowning            205
Drugs               192
Explosives          149
Poison               78
Fall                 24
Name: Weapon, dtype: int64

In [88]:
unsolved_victims = unsolved.loc[:,'Victim Sex':'Victim Ethnicity']
unsolved_victims.head(10)

Victim Sex  Victim Age                    Victim Race Victim Ethnicity
2      Female          30  Native American/Alaska Native          Unknown
4      Female          30  Native American/Alaska Native          Unknown
8        Male          32                          White          Unknown
10       Male          36  Native American/Alaska Native          Unknown
14       Male          48                          White          Unknown
26       Male          27                          Black          Unknown
32       Male          25                          White          Unknown
33       Male          24  Native American/Alaska Native          Unknown
34       Male          33                          White          Unknown
38     Female          22                          White          Unknown

### Victim Sex

In [89]:
unsolved_victims['Victim Sex'].value_counts()

Male       156552
Female      33076
Unknown       653
Name: Victim Sex, dtype: int64

### Victim Age

In [90]:
# Check if any victim ages are less than 1.  Is 0 indicating a baby under 1 years old or unknown age???
check_victim_ages = unsolved_victims[unsolved_victims['Victim Age'] < 1]
print(check_victim_ages.shape)
check_victim_ages.head()

(1025, 4)


Victim Sex  Victim Age Victim Race Victim Ethnicity
604      Female           0       White          Unknown
1511     Female           0       White     Not Hispanic
1958       Male           0       White     Not Hispanic
2945       Male           0       White     Not Hispanic
3491     Female           0       White          Unknown

In [91]:
# Looking at one of these rows, there's really not much information to tell me anything about this age 0 victim
homicide.iloc[1958]

City                                Los Angeles
State                                California
Year                                       1980
Month                                  November
Incident                                      1
Crime Type               Murder or Manslaughter
Crime Solved                                 No
Victim Sex                                 Male
Victim Age                                    0
Victim Race                               White
Victim Ethnicity                   Not Hispanic
Perpetrator Sex                         Unknown
Perpetrator Age                               0
Perpetrator Race                        Unknown
Perpetrator Ethnicity                   Unknown
Relationship                            Unknown
Weapon                             Blunt Object
Victim Count                                  0
Perpetrator Count                             0
Record Source                               FBI
Name: 1958, dtype: object

### Victim Race

In [92]:
unsolved_victims['Victim Race'].value_counts()

Black                            101124
White                             82236
Unknown                            3084
Asian/Pacific Islander             2917
Native American/Alaska Native       920
Name: Victim Race, dtype: int64

### Victim Ethnicity

In [93]:
unsolved_victims['Victim Ethnicity'].value_counts()

Unknown         113271
Not Hispanic     51918
Hispanic         25092
Name: Victim Ethnicity, dtype: int64

## Data Preprocessing

In [451]:
solved = homicide[homicide['Crime Solved'] == 'Yes']
solved.head()

City   State  Year    Month  Incident              Crime Type  \
0  Anchorage  Alaska  1980  January         1  Murder or Manslaughter   
1  Anchorage  Alaska  1980    March         1  Murder or Manslaughter   
3  Anchorage  Alaska  1980    April         1  Murder or Manslaughter   
5  Anchorage  Alaska  1980      May         1  Murder or Manslaughter   
6  Anchorage  Alaska  1980      May         2  Murder or Manslaughter   

  Crime Solved Victim Sex  Victim Age                    Victim Race  \
0          Yes       Male          14  Native American/Alaska Native   
1          Yes       Male          43                          White   
3          Yes       Male          43                          White   
5          Yes       Male          30                          White   
6          Yes     Female          42  Native American/Alaska Native   

  Victim Ethnicity Perpetrator Sex  Perpetrator Age  \
0          Unknown            Male               15   
1          Unknown            Male               42   
3          Unknown            Male               42   
5          Unknown            Male               36   
6          Unknown            Male               27   

                Perpetrator Race Perpetrator Ethnicity  Relationship  \
0  Native American/Alaska Native               Unknown  Acquaintance   
1                          White               Unknown  Acquaintance   
3                          White               Unknown  Acquaintance   
5                          White               Unknown  Acquaintance   
6                          Black               Unknown          Wife   

          Weapon  Victim Count  Perpetrator Count Record Source  
0   Blunt Object             0                  0           FBI  
1  Strangulation             0                  0           FBI  
3  Strangulation             0                  0           FBI  
5          Rifle             0                  0           FBI  
6          Knife             0                  0           FBI

In [452]:
# Replace 'Unknown' with NaN, then drop those rows
solved = solved.replace('Unknown', np.nan)
solved.isnull().values.any()

True

In [453]:
solved.dropna(axis=0, how='any', inplace=True)
solved.isnull().values.any()

False

In [454]:
solved.shape

(155853, 20)

In [455]:
# There's still plenty of rows to work with after dropping all the Unknowns!
solved.head()

City    State  Year      Month  Incident                  Crime Type  \
44     Juneau   Alaska  1980   November         4  Manslaughter by Negligence   
52  Jefferson  Alabama  1980       July         1      Murder or Manslaughter   
53  Jefferson  Alabama  1980       July         2      Murder or Manslaughter   
54  Jefferson  Alabama  1980       July         3      Murder or Manslaughter   
58  Jefferson  Alabama  1980  September        41      Murder or Manslaughter   

   Crime Solved Victim Sex  Victim Age                    Victim Race  \
44          Yes     Female          26  Native American/Alaska Native   
52          Yes       Male          23                          Black   
53          Yes       Male          42                          White   
54          Yes       Male          33                          Black   
58          Yes       Male          46                          White   

   Victim Ethnicity Perpetrator Sex  Perpetrator Age  \
44     Not Hispanic            Male               26   
52     Not Hispanic            Male               32   
53     Not Hispanic          Female               40   
54     Not Hispanic            Male               26   
58     Not Hispanic            Male               22   

                 Perpetrator Race Perpetrator Ethnicity  Relationship  \
44  Native American/Alaska Native          Not Hispanic          Wife   
52                          Black          Not Hispanic  Acquaintance   
53                          White          Not Hispanic       Husband   
54                          Black          Not Hispanic  Acquaintance   
58                          White          Not Hispanic        Father   

      Weapon  Victim Count  Perpetrator Count Record Source  
44  Drowning             0                  0           FBI  
52   Handgun             0                  0           FBI  
53   Handgun             0                  0           FBI  
54   Handgun             0                  1           FBI  
58     Rifle             0                  0           FBI

In [456]:
# Drop the following columns: City State, Year, Month, Incident, Crime Type, Crime Solved, Victim Count, Perpetrator Count, Record Source
solved = solved.loc[:,'Victim Sex':'Weapon']
solved.head()

Victim Sex  Victim Age                    Victim Race Victim Ethnicity  \
44     Female          26  Native American/Alaska Native     Not Hispanic   
52       Male          23                          Black     Not Hispanic   
53       Male          42                          White     Not Hispanic   
54       Male          33                          Black     Not Hispanic   
58       Male          46                          White     Not Hispanic   

   Perpetrator Sex  Perpetrator Age               Perpetrator Race  \
44            Male               26  Native American/Alaska Native   
52            Male               32                          Black   
53          Female               40                          White   
54            Male               26                          Black   
58            Male               22                          White   

   Perpetrator Ethnicity  Relationship    Weapon  
44          Not Hispanic          Wife  Drowning  
52          Not Hispanic  Acquaintance   Handgun  
53          Not Hispanic       Husband   Handgun  
54          Not Hispanic  Acquaintance   Handgun  
58          Not Hispanic        Father     Rifle

In [457]:
solved['Relationship'].value_counts()

Acquaintance            50723
Stranger                42520
Friend                  11540
Wife                     9878
Girlfriend               6460
Husband                  4348
Family                   3808
Son                      3500
Neighbor                 3078
Boyfriend                2882
Daughter                 2643
Brother                  2470
In-Law                   2023
Father                   1815
Mother                   1675
Common-Law Wife          1195
Common-Law Husband       1021
Ex-Wife                   855
Stepfather                659
Boyfriend/Girlfriend      639
Stepson                   540
Sister                    479
Stepdaughter              312
Ex-Husband                303
Employer                  241
Employee                  159
Stepmother                 87
Name: Relationship, dtype: int64

In [458]:
# Label Encoding
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [459]:
# Function for label encoding multiple columns
le_variables = {}
def label_encoder(df, df_columns):
    for column in df_columns:
        le = LabelEncoder()
        labels = le.fit_transform(df[column])
        le_variables[column] = le
        df['LE_' + column] = labels

In [460]:
# Label encode all categorical columns 
columns_to_exclude = ['Victim Age', 'Perpetrator Age']
columns_to_encode = [column for column in solved.columns if column not in columns_to_exclude]
label_encoder(solved, columns_to_encode)
solved.head()

Victim Sex  Victim Age                    Victim Race Victim Ethnicity  \
44     Female          26  Native American/Alaska Native     Not Hispanic   
52       Male          23                          Black     Not Hispanic   
53       Male          42                          White     Not Hispanic   
54       Male          33                          Black     Not Hispanic   
58       Male          46                          White     Not Hispanic   

   Perpetrator Sex  Perpetrator Age               Perpetrator Race  \
44            Male               26  Native American/Alaska Native   
52            Male               32                          Black   
53          Female               40                          White   
54            Male               26                          Black   
58            Male               22                          White   

   Perpetrator Ethnicity  Relationship    Weapon  LE_Victim Sex  \
44          Not Hispanic          Wife  Drowning              0   
52          Not Hispanic  Acquaintance   Handgun              1   
53          Not Hispanic       Husband   Handgun              1   
54          Not Hispanic  Acquaintance   Handgun              1   
58          Not Hispanic        Father     Rifle              1   

    LE_Victim Race  LE_Victim Ethnicity  LE_Perpetrator Sex  \
44               2                    1                   1   
52               1                    1                   1   
53               3                    1                   0   
54               1                    1                   1   
58               3                    1                   1   

    LE_Perpetrator Race  LE_Perpetrator Ethnicity  LE_Relationship  LE_Weapon  
44                    2                         1               26          1  
52                    1                         1                0          8  
53                    3                         1               15          8  
54                    1                         1                0          8  
58                    3                         1               12         11

In [461]:
# Function for one hot encoding
def oh_encoder(df, df_columns):
    new_df = df
    for column in df_columns:
        ohe = OneHotEncoder()
        feature_array = ohe.fit_transform(df[[column]]).toarray()
        
        # Prepend 'victim' or 'perpetrator' if column name starts with either
        # Required to differentiate victim and perpetrator sex, race, ethnicity once encoded
        column_names = list(le_variables[column[3:]].classes_)
        if column.startswith('LE_Victim'):
            column_names = ['Victim ' + name for name in column_names]
        elif column.startswith('LE_Perpetrator'):
            column_names = ['Perpetrator ' + name for name in column_names]
        
        # Concatenate new one-hot encoded dfs with original df
        encoded_df = pd.DataFrame(feature_array, columns = column_names)
        new_df = pd.concat([new_df.reset_index(drop=True), encoded_df.reset_index(drop=True)], axis=1)
    return new_df

In [462]:
columns_to_oh_encode = [column for column in solved.columns if column.startswith('LE_')]
solved = oh_encoder(solved, columns_to_oh_encode)
solved.head()

Victim Sex  Victim Age                    Victim Race Victim Ethnicity  \
0     Female          26  Native American/Alaska Native     Not Hispanic   
1       Male          23                          Black     Not Hispanic   
2       Male          42                          White     Not Hispanic   
3       Male          33                          Black     Not Hispanic   
4       Male          46                          White     Not Hispanic   

  Perpetrator Sex  Perpetrator Age               Perpetrator Race  \
0            Male               26  Native American/Alaska Native   
1            Male               32                          Black   
2          Female               40                          White   
3            Male               26                          Black   
4            Male               22                          White   

  Perpetrator Ethnicity  Relationship    Weapon     ...       Fire  Firearm  \
0          Not Hispanic          Wife  Drowning     ...        0.0      0.0   
1          Not Hispanic  Acquaintance   Handgun     ...        0.0      0.0   
2          Not Hispanic       Husband   Handgun     ...        0.0      0.0   
3          Not Hispanic  Acquaintance   Handgun     ...        0.0      0.0   
4          Not Hispanic        Father     Rifle     ...        0.0      0.0   

   Gun  Handgun  Knife  Poison  Rifle  Shotgun  Strangulation  Suffocation  
0  0.0      0.0    0.0     0.0    0.0      0.0            0.0          0.0  
1  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
2  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
3  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
4  0.0      0.0    0.0     0.0    1.0      0.0            0.0          0.0  

[5 rows x 76 columns]

In [463]:
solved.columns

Index(['Victim Sex', 'Victim Age', 'Victim Race', 'Victim Ethnicity',
       'Perpetrator Sex', 'Perpetrator Age', 'Perpetrator Race',
       'Perpetrator Ethnicity', 'Relationship', 'Weapon', 'LE_Victim Sex',
       'LE_Victim Race', 'LE_Victim Ethnicity', 'LE_Perpetrator Sex',
       'LE_Perpetrator Race', 'LE_Perpetrator Ethnicity', 'LE_Relationship',
       'LE_Weapon', 'Victim Female', 'Victim Male',
       'Victim Asian/Pacific Islander', 'Victim Black',
       'Victim Native American/Alaska Native', 'Victim White',
       'Victim Hispanic', 'Victim Not Hispanic', 'Perpetrator Female',
       'Perpetrator Male', 'Perpetrator Asian/Pacific Islander',
       'Perpetrator Black', 'Perpetrator Native American/Alaska Native',
       'Perpetrator White', 'Perpetrator Hispanic', 'Perpetrator Not Hispanic',
       'Acquaintance', 'Boyfriend', 'Boyfriend/Girlfriend', 'Brother',
       'Common-Law Husband', 'Common-Law Wife', 'Daughter', 'Employee',
       'Employer', 'Ex-Husband', 'Ex-Wif

# Supervised Learning

## K-Nearest Neighbors

The following fits the solved dataframe with all categorcal columns encoded into a K-Nearest Neighbors algorithm.

In [312]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

In [480]:
# Reset index to prevent columns being converted to NaN values
X1 = solved[['Victim Age']].reset_index(drop=True)
X2 = solved.loc[:,'Victim Female':'Victim Not Hispanic'].reset_index(drop=True)
X3 = solved.loc[:,'Blunt Object':]
X = pd.concat([X1, X2, X3], axis=1)
display(X.head())

y = solved.loc[:,'Perpetrator Female':'Wife']
y.head()

Victim Age  Victim Female  Victim Male  Victim Asian/Pacific Islander  \
0          26            1.0          0.0                            0.0   
1          23            0.0          1.0                            0.0   
2          42            0.0          1.0                            0.0   
3          33            0.0          1.0                            0.0   
4          46            0.0          1.0                            0.0   

   Victim Black  Victim Native American/Alaska Native  Victim White  \
0           0.0                                   1.0           0.0   
1           1.0                                   0.0           0.0   
2           0.0                                   0.0           1.0   
3           1.0                                   0.0           0.0   
4           0.0                                   0.0           1.0   

   Victim Hispanic  Victim Not Hispanic  Blunt Object     ...       Fire  \
0              0.0                  1.0           0.0     ...        0.0   
1              0.0                  1.0           0.0     ...        0.0   
2              0.0                  1.0           0.0     ...        0.0   
3              0.0                  1.0           0.0     ...        0.0   
4              0.0                  1.0           0.0     ...        0.0   

   Firearm  Gun  Handgun  Knife  Poison  Rifle  Shotgun  Strangulation  \
0      0.0  0.0      0.0    0.0     0.0    0.0      0.0            0.0   
1      0.0  0.0      1.0    0.0     0.0    0.0      0.0            0.0   
2      0.0  0.0      1.0    0.0     0.0    0.0      0.0            0.0   
3      0.0  0.0      1.0    0.0     0.0    0.0      0.0            0.0   
4      0.0  0.0      0.0    0.0     0.0    1.0      0.0            0.0   

   Suffocation  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  

[5 rows x 24 columns]

Perpetrator Female  Perpetrator Male  Perpetrator Asian/Pacific Islander  \
0                 0.0               1.0                                 0.0   
1                 0.0               1.0                                 0.0   
2                 1.0               0.0                                 0.0   
3                 0.0               1.0                                 0.0   
4                 0.0               1.0                                 0.0   

   Perpetrator Black  Perpetrator Native American/Alaska Native  \
0                0.0                                        1.0   
1                1.0                                        0.0   
2                0.0                                        0.0   
3                1.0                                        0.0   
4                0.0                                        0.0   

   Perpetrator White  Perpetrator Hispanic  Perpetrator Not Hispanic  \
0                0.0                   0.0                       1.0   
1                0.0                   0.0                       1.0   
2                1.0                   0.0                       1.0   
3                0.0                   0.0                       1.0   
4                1.0                   0.0                       1.0   

   Acquaintance  Boyfriend  ...   Mother  Neighbor  Sister  Son  Stepdaughter  \
0           0.0        0.0  ...      0.0       0.0     0.0  0.0           0.0   
1           1.0        0.0  ...      0.0       0.0     0.0  0.0           0.0   
2           0.0        0.0  ...      0.0       0.0     0.0  0.0           0.0   
3           1.0        0.0  ...      0.0       0.0     0.0  0.0           0.0   
4           0.0        0.0  ...      0.0       0.0     0.0  0.0           0.0   

   Stepfather  Stepmother  Stepson  Stranger  Wife  
0         0.0         0.0      0.0       0.0   1.0  
1         0.0         0.0      0.0       0.0   0.0  
2         0.0         0.0      0.0       0.0   0.0  
3         0.0         0.0      0.0       0.0   0.0  
4         0.0         0.0      0.0       0.0   0.0  

[5 rows x 35 columns]

In [481]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.3,random_state = 42)

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

In [482]:
knn.score(X_test, y_test)

0.13469928993070407

WOW, that is a really terrible accuracy score!!!!
Accuracy: 13.47%

## K-Nearest Neighbors - Take 2

Let's try that again, but this time, the ages will be categorized instead

In [471]:
solved2 = solved.copy()
solved2.head()

Victim Sex  Victim Age                    Victim Race Victim Ethnicity  \
0     Female          26  Native American/Alaska Native     Not Hispanic   
1       Male          23                          Black     Not Hispanic   
2       Male          42                          White     Not Hispanic   
3       Male          33                          Black     Not Hispanic   
4       Male          46                          White     Not Hispanic   

  Perpetrator Sex  Perpetrator Age               Perpetrator Race  \
0            Male               26  Native American/Alaska Native   
1            Male               32                          Black   
2          Female               40                          White   
3            Male               26                          Black   
4            Male               22                          White   

  Perpetrator Ethnicity  Relationship    Weapon     ...       Fire  Firearm  \
0          Not Hispanic          Wife  Drowning     ...        0.0      0.0   
1          Not Hispanic  Acquaintance   Handgun     ...        0.0      0.0   
2          Not Hispanic       Husband   Handgun     ...        0.0      0.0   
3          Not Hispanic  Acquaintance   Handgun     ...        0.0      0.0   
4          Not Hispanic        Father     Rifle     ...        0.0      0.0   

   Gun  Handgun  Knife  Poison  Rifle  Shotgun  Strangulation  Suffocation  
0  0.0      0.0    0.0     0.0    0.0      0.0            0.0          0.0  
1  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
2  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
3  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
4  0.0      0.0    0.0     0.0    1.0      0.0            0.0          0.0  

[5 rows x 76 columns]

In [472]:
# Age ranges: < 15, 15-24, 25-34, 35-44, 45-54, 55-64, 65+
def categorize_age(age):
    if age < 15:
        return '<15'
    elif age >= 15 and age <= 24:
        return '15-24'
    elif age >= 25 and age <= 34:
        return '25-34'
    elif age >= 35 and age <= 44:
        return '35-44'
    elif age >= 45 and age <= 54:
        return '45-54'
    elif age >= 55 and age <= 64:
        return '55-64'
    elif age >= 65:
        return '65+'

In [473]:
# solved2['Victim Age'].apply(str)
# solved2['Perpetrator Age'].apply(str)

solved2['Victim Age'] = solved2['Victim Age'].apply(categorize_age)
solved2['Perpetrator Age'] = solved2['Perpetrator Age'].apply(categorize_age)

solved2.head()

Victim Sex Victim Age                    Victim Race Victim Ethnicity  \
0     Female      25-34  Native American/Alaska Native     Not Hispanic   
1       Male      15-24                          Black     Not Hispanic   
2       Male      35-44                          White     Not Hispanic   
3       Male      25-34                          Black     Not Hispanic   
4       Male      45-54                          White     Not Hispanic   

  Perpetrator Sex Perpetrator Age               Perpetrator Race  \
0            Male           25-34  Native American/Alaska Native   
1            Male           25-34                          Black   
2          Female           35-44                          White   
3            Male           25-34                          Black   
4            Male           15-24                          White   

  Perpetrator Ethnicity  Relationship    Weapon     ...       Fire  Firearm  \
0          Not Hispanic          Wife  Drowning     ...        0.0      0.0   
1          Not Hispanic  Acquaintance   Handgun     ...        0.0      0.0   
2          Not Hispanic       Husband   Handgun     ...        0.0      0.0   
3          Not Hispanic  Acquaintance   Handgun     ...        0.0      0.0   
4          Not Hispanic        Father     Rifle     ...        0.0      0.0   

   Gun  Handgun  Knife  Poison  Rifle  Shotgun  Strangulation  Suffocation  
0  0.0      0.0    0.0     0.0    0.0      0.0            0.0          0.0  
1  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
2  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
3  0.0      1.0    0.0     0.0    0.0      0.0            0.0          0.0  
4  0.0      0.0    0.0     0.0    1.0      0.0            0.0          0.0  

[5 rows x 76 columns]

In [474]:
label_encoder(solved2, ['Victim Age', 'Perpetrator Age'])
solved2 = oh_encoder(solved2, ['LE_Victim Age', 'LE_Perpetrator Age'])
print(solved2.columns)
solved2.head()

Index(['Victim Sex', 'Victim Age', 'Victim Race', 'Victim Ethnicity',
       'Perpetrator Sex', 'Perpetrator Age', 'Perpetrator Race',
       'Perpetrator Ethnicity', 'Relationship', 'Weapon', 'LE_Victim Sex',
       'LE_Victim Race', 'LE_Victim Ethnicity', 'LE_Perpetrator Sex',
       'LE_Perpetrator Race', 'LE_Perpetrator Ethnicity', 'LE_Relationship',
       'LE_Weapon', 'Victim Female', 'Victim Male',
       'Victim Asian/Pacific Islander', 'Victim Black',
       'Victim Native American/Alaska Native', 'Victim White',
       'Victim Hispanic', 'Victim Not Hispanic', 'Perpetrator Female',
       'Perpetrator Male', 'Perpetrator Asian/Pacific Islander',
       'Perpetrator Black', 'Perpetrator Native American/Alaska Native',
       'Perpetrator White', 'Perpetrator Hispanic', 'Perpetrator Not Hispanic',
       'Acquaintance', 'Boyfriend', 'Boyfriend/Girlfriend', 'Brother',
       'Common-Law Husband', 'Common-Law Wife', 'Daughter', 'Employee',
       'Employer', 'Ex-Husband', 'Ex-Wif

Victim Sex Victim Age                    Victim Race Victim Ethnicity  \
0     Female      25-34  Native American/Alaska Native     Not Hispanic   
1       Male      15-24                          Black     Not Hispanic   
2       Male      35-44                          White     Not Hispanic   
3       Male      25-34                          Black     Not Hispanic   
4       Male      45-54                          White     Not Hispanic   

  Perpetrator Sex Perpetrator Age               Perpetrator Race  \
0            Male           25-34  Native American/Alaska Native   
1            Male           25-34                          Black   
2          Female           35-44                          White   
3            Male           25-34                          Black   
4            Male           15-24                          White   

  Perpetrator Ethnicity  Relationship    Weapon       ...         \
0          Not Hispanic          Wife  Drowning       ...          
1          Not Hispanic  Acquaintance   Handgun       ...          
2          Not Hispanic       Husband   Handgun       ...          
3          Not Hispanic  Acquaintance   Handgun       ...          
4          Not Hispanic        Father     Rifle       ...          

   Victim 55-64  Victim 65+  Victim <15  Perpetrator 15-24  Perpetrator 25-34  \
0           0.0         0.0         0.0                0.0                1.0   
1           0.0         0.0         0.0                0.0                1.0   
2           0.0         0.0         0.0                0.0                0.0   
3           0.0         0.0         0.0                0.0                1.0   
4           0.0         0.0         0.0                1.0                0.0   

   Perpetrator 35-44  Perpetrator 45-54  Perpetrator 55-64  Perpetrator 65+  \
0                0.0                0.0                0.0              0.0   
1                0.0                0.0                0.0              0.0   
2                1.0                0.0                0.0              0.0   
3                0.0                0.0                0.0              0.0   
4                0.0                0.0                0.0              0.0   

   Perpetrator <15  
0              0.0  
1              0.0  
2              0.0  
3              0.0  
4              0.0  

[5 rows x 92 columns]

In [485]:
victim_1 = solved2.loc[:,'Victim 15-24':'Victim <15'].reset_index(drop=True)
victim_2 = solved2.loc[:,'Victim Female':'Victim Not Hispanic'].reset_index(drop=True)
victim_3 = solved2.loc[:,'Blunt Object':'Suffocation']
X2 = pd.concat([victim_1, victim_2, victim_3], axis=1)
display(X2.head())

perp_1 = solved2.loc[:,'Perpetrator Female':'Wife']
perp_2 = solved2.loc[:, 'Perpetrator 15-24':]
y2 = pd.concat([perp_1, perp_2], axis=1)
y2.head()

Victim 15-24  Victim 25-34  Victim 35-44  Victim 45-54  Victim 55-64  \
0           0.0           1.0           0.0           0.0           0.0   
1           1.0           0.0           0.0           0.0           0.0   
2           0.0           0.0           1.0           0.0           0.0   
3           0.0           1.0           0.0           0.0           0.0   
4           0.0           0.0           0.0           1.0           0.0   

   Victim 65+  Victim <15  Victim Female  Victim Male  \
0         0.0         0.0            1.0          0.0   
1         0.0         0.0            0.0          1.0   
2         0.0         0.0            0.0          1.0   
3         0.0         0.0            0.0          1.0   
4         0.0         0.0            0.0          1.0   

   Victim Asian/Pacific Islander     ...       Fire  Firearm  Gun  Handgun  \
0                            0.0     ...        0.0      0.0  0.0      0.0   
1                            0.0     ...        0.0      0.0  0.0      1.0   
2                            0.0     ...        0.0      0.0  0.0      1.0   
3                            0.0     ...        0.0      0.0  0.0      1.0   
4                            0.0     ...        0.0      0.0  0.0      0.0   

   Knife  Poison  Rifle  Shotgun  Strangulation  Suffocation  
0    0.0     0.0    0.0      0.0            0.0          0.0  
1    0.0     0.0    0.0      0.0            0.0          0.0  
2    0.0     0.0    0.0      0.0            0.0          0.0  
3    0.0     0.0    0.0      0.0            0.0          0.0  
4    0.0     0.0    1.0      0.0            0.0          0.0  

[5 rows x 30 columns]

Perpetrator Female  Perpetrator Male  Perpetrator Asian/Pacific Islander  \
0                 0.0               1.0                                 0.0   
1                 0.0               1.0                                 0.0   
2                 1.0               0.0                                 0.0   
3                 0.0               1.0                                 0.0   
4                 0.0               1.0                                 0.0   

   Perpetrator Black  Perpetrator Native American/Alaska Native  \
0                0.0                                        1.0   
1                1.0                                        0.0   
2                0.0                                        0.0   
3                1.0                                        0.0   
4                0.0                                        0.0   

   Perpetrator White  Perpetrator Hispanic  Perpetrator Not Hispanic  \
0                0.0                   0.0                       1.0   
1                0.0                   0.0                       1.0   
2                1.0                   0.0                       1.0   
3                0.0                   0.0                       1.0   
4                1.0                   0.0                       1.0   

   Acquaintance  Boyfriend       ...         Stepson  Stranger  Wife  \
0           0.0        0.0       ...             0.0       0.0   1.0   
1           1.0        0.0       ...             0.0       0.0   0.0   
2           0.0        0.0       ...             0.0       0.0   0.0   
3           1.0        0.0       ...             0.0       0.0   0.0   
4           0.0        0.0       ...             0.0       0.0   0.0   

   Perpetrator 15-24  Perpetrator 25-34  Perpetrator 35-44  Perpetrator 45-54  \
0                0.0                1.0                0.0                0.0   
1                0.0                1.0                0.0                0.0   
2                0.0                0.0                1.0                0.0   
3                0.0                1.0                0.0                0.0   
4                1.0                0.0                0.0                0.0   

   Perpetrator 55-64  Perpetrator 65+  Perpetrator <15  
0                0.0              0.0              0.0  
1                0.0              0.0              0.0  
2                0.0              0.0              0.0  
3                0.0              0.0              0.0  
4                0.0              0.0              0.0  

[5 rows x 42 columns]

In [488]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y2, test_size = 0.3, random_state = 42)

knn2 = KNeighborsClassifier(n_neighbors=15)
knn2.fit(X_train2, y_train2)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=15, p=2,
           weights='uniform')

In [489]:
knn2.score(X_test2, y_test2)

0.043096073231243051